src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -660,101 +661,130 @@ public long GetNumberOfCounterEntries(DocumentsOperationContext context)
         }
 
-        public void UpdateDocumentCounters(DocumentsOperationContext context, BlittableJsonReaderObject doc, string docId, BlittableJsonReaderObject metadata, List<CounterOperation> countersOperations)
+        public void UpdateDocumentCounters(DocumentsOperationContext context, BlittableJsonReaderObject doc, string docId, List<CounterOperation> countersOperations)
         {
-            List<string> updates = null;
-            if (metadata.TryGet(Constants.Documents.Metadata.Counters, out BlittableJsonReaderArray counters))
+            BlittableJsonReaderArray metadataCounters = null;
+            var initCounterListSize = countersOperations.Count;
+            if (doc.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata) &&
+                metadata.TryGet(Constants.Documents.Metadata.Counters, out metadataCounters))
             {
-                foreach (var operation in countersOperations)
+                initCounterListSize += metadataCounters.Length;
+            }
+
+            //Todo Maybe should not read from the table if there is metadata on document
+            List<string> counters = null;
+            foreach (var s in GetCountersForDocument(context, docId))
+            {
+                if (null == counters)
                 {
-                    // we need to check the updates to avoid inserting duplicate counter names
-                    int loc = updates?.BinarySearch(operation.CounterName, StringComparer.OrdinalIgnoreCase) ??
-                              counters.BinarySearch(operation.CounterName, StringComparison.OrdinalIgnoreCase);
+                    counters = new List<string>(initCounterListSize);
+                }
+                counters.Add(s);
+            }
 
-                    switch (operation.Type)
-                    {
-                        case CounterOperationType.Increment:
-                        case CounterOperationType.Put:
-                            if (loc < 0)
-                            {
-                                CreateUpdatesIfNeeded();
-                                updates.Insert(~loc, operation.CounterName);
-                            }
-
-                            break;
-                        case CounterOperationType.Delete:
-                            if (loc >= 0)
-                            {
-                                CreateUpdatesIfNeeded();
-                                updates.RemoveAt(loc);
-                            }
-                            break;
-                        case CounterOperationType.None:
-                        case CounterOperationType.Get:
-                            break;
-                        default:
-                            throw new ArgumentOutOfRangeException($"Unknown value {operation.Type}");
-                    }
+            //Todo To consider if can count on that this already in order
+            counters?.Sort(StringComparer.OrdinalIgnoreCase);
```

Comment:
When reading from the table, yes, you can rely on it being sorted based on the index you are scanning